File: block/python/ansi_x9.17.py
#!/usr/bin/python

import argparse
import struct
import time
from Crypto.Cipher import AES
from Crypto.Protocol import KDF

parser = argparse.ArgumentParser(description='ANSI x9.17 DRBG')
parser.add_argument('-k','--key', help='AES key.')
parser.add_argument('-s','--seed', help='Starting seed.')
parser.add_argument('-n','--numbers', help='Quantity of random numbers.')
args = parser.parse_args()

def sxor(s1, s2):
    return ''.join(chr(ord(a)^ord(b)) for a,b in zip(s1,s2))

salt = '8bbb35cb1626f7c9ef291b29261acbf5' # for PBKDF2
key = KDF.PBKDF2('11cf0edfc106350f9d81abe1538b3f20', salt, 16, 0)
seed = KDF.PBKDF2('11462794b14c20f6b0896bac3a921485', salt, 16, 0)

if args.key:
    key = KDF.PBKDF2(bytes(args.key), salt, 16, 0)

if args.seed:
    seed = KDF.PBKDF2(bytes(args.seed), salt, 16, 0)

if args.numbers:
    number = int(args.numbers)
else:
    number = 1

aes = AES.new(key)

# the actual ANSI X9.17 algorithm
for i in range(0, number):
    date = repr(time.time())
    temp = aes.encrypt(date.zfill(32))
    out = aes.encrypt(sxor(seed,temp))
    seed = aes.encrypt(sxor(out,temp))
    res = struct.unpack('QQ', out)
    print res[0]+res[1]*2**64

File: block/perl/ansi_x9.17.pl
#!/usr/bin/perl
use strict;

use Crypt::PBKDF2;      # requires libcrypt-pbkdf2-perl
use Crypt::Rijndael;
use Crypt::Digest::RIPEMD160;
use Digest::SHA;
use Getopt::Long qw(GetOptions);
use Time::HiRes;

my $b;  # display raw binary output
my $k;  # user-supplied key
my $s;  # user-supplied seed
my $n;  # user-supplied rounds
my $h;  # display usage

my $number;
my $pbkdf2;

my $res = GetOptions(
    "b|binary"    => \$b,
    "k|key=s"     => \$k,
    "s|seed=s"    => \$s,
    "n|numbers=i" => \$n,
    "h|help"      => \$h,
);

open(FILE, '/proc/interrupts');
my $data = join('',<FILE>);
close(FILE);

my $sha = Digest::SHA::sha1($data);
my $ripemd = Crypt::Digest::RIPEMD160::ripemd160($data);

$pbkdf2 = Crypt::PBKDF2->new(
    hash_class  => 'HMACSHA1',
    # 0 iterations guarantees high performance 16-byte output
    # the security rests on the entropy of $key, not iterating pbkdf2
    iterations  => 0,
    output_len  => 16,
    salt_len    => 4,
);

if ($h) {
    print "usage: ansi_x9.17.prl [-h] [-k KEY] [-s SEED] [-n NUMBERS]

ANSI x9.17 DRBG

optional arguments:
  -h, --help            Show this help message and exit.
  -b, --binary          Raw binary output.
  -k KEY, --key KEY     AES key.
  -s SEED, --seed SEED  Starting seed.
  -n NUMBERS, --numbers NUMBERS
                        Quantity of random numbers.
";
    exit;
}

my $salt = $sha^$ripemd;
my $key = $pbkdf2->PBKDF2($sha, $salt);
my $seed = $pbkdf2->PBKDF2($ripemd, $salt);

$key = $pbkdf2->PBKDF2($k, $salt) if ($k);
$seed = $pbkdf2->PBKDF2($s, $salt) if ($s);
$number = ($n ? int($n) : 1);

my $rijndael = Crypt::Rijndael->new($key);

# the actual ANSI X9.17 algorithm
for (my $i=0; $i<$number; $i++) {
    my $date = $pbkdf2->PBKDF2(Time::HiRes::time(), $salt);
    my $temp = $rijndael->encrypt($date);
    my $out = $rijndael->encrypt($seed ^ $temp);
    $seed = $rijndael->encrypt($out ^ $temp);
    if ($b) { print $out; }
    else { print unpack("QQ", $out), "\n"; }
}
